### PR TITLE
Refs [TASK][26258] Create Datasource and Repository

### PR DESCRIPTION
**Pull request by quockhanhng**
## Related Tickets
- [#26258](https://edu-redmine.sun-asterisk.vn/issues/26258)

## WHAT
- Create Datasource and Repository

## HOW
- I define required functions in the DataSource interface and implement them in the LocalDataSource class. In these implemented functions, I use AsyncTask to execute data through a DAO instance 
- In the Repository class, because there's only local data, I also implement directly from DataSource.Local without creating a Repository interface

## WHY
- Organize code such that the presenter class doesn't need to care about where data comes from. It only cares about how to request data and what it gets back.

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Reviewer1 (KhanhNQ) | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/TaskDAOImpl.kt [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            cursor.getString(cursor.getColumnIndex("TITLE")),
+            cursor.getString(cursor.getColumnIndex("DESCRIPTION")),
+            cursor.getString(cursor.getColumnIndex("TIME_REMINDER")),
+            cursor.getString(cursor.getColumnIndex("IS_IMPORTANT")) == "true",
+            cursor.getString(cursor.getColumnIndex("IS_IN_MY_DAY")) == "true",
+            cursor.getString(cursor.getColumnIndex("COLOR")),
+            cursor.getInt(cursor.getColumnIndex("STATUS")),
+            cursor.getString(cursor.getColumnIndex("TIME_CREATED"))
+        )
+    }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/TaskDAOImpl.kt]: Chuyển về trong class Task nhé 
```suggestion
    private fun getTaskFromCursor(cursor: Cursor) = Task(cursor) 
```

File: app/src/main/java/com/sunasterisk/getitdone/data/repository/TaskListRepository.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.sunasterisk.getitdone.data.repository
+
+import com.sunasterisk.getitdone.data.OnLoadedCallback
+import com.sunasterisk.getitdone.data.model.TaskList
+import com.sunasterisk.getitdone.data.source.TaskListDataSource
+
+class TaskListRepository private constructor(private val localDataSource: TaskListDataSource.Local) :
+    TaskListDataSource.Local {
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/getitdone/data/repository/TaskListRepository.kt]:
```suggestion
class TaskListRepository private constructor(
    private val localDataSource: TaskListDataSource.Local
) : TaskListDataSource.Local {
```

File: app/src/main/java/com/sunasterisk/getitdone/data/repository/TaskRepository.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
+package com.sunasterisk.getitdone.data.repository
+
+import com.sunasterisk.getitdone.data.OnLoadedCallback
+import com.sunasterisk.getitdone.data.model.Task
+import com.sunasterisk.getitdone.data.source.TaskDataSource
+
+class TaskRepository private constructor(private val localDataSource: TaskDataSource.Local) :
+    TaskDataSource.Local {
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/getitdone/data/repository/TaskRepository.kt]:
```suggestion
class TaskRepository private constructor(
     private val localDataSource: TaskDataSource.Local
 ) : TaskDataSource.Local {
```

File: app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/TaskDAOImpl.kt (hunk 1 above)
Comment (quockhanhng) [reply app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/TaskDAOImpl.kt]: vâng ạ

File: app/src/main/java/com/sunasterisk/getitdone/data/repository/TaskListRepository.kt (hunk 2 above)
Comment (quockhanhng) [reply app/src/main/java/com/sunasterisk/getitdone/data/repository/TaskListRepository.kt]: fixed :+1:

File: app/src/main/java/com/sunasterisk/getitdone/data/repository/TaskRepository.kt (hunk 3 above)
Comment (quockhanhng) [reply app/src/main/java/com/sunasterisk/getitdone/data/repository/TaskRepository.kt]: fixed :+1:

**Review by datdt-1935 — APPROVED** [review]

**Pull request merged**